# Predicting Average House Prices in Madrid Neighbourhoods Using Socioeconomic Indicators

## Overall Project Aim:
To develop and evaluate a machine learning model that uses 2025 socioeconomic indicators to predict average housing prices across Madrid neighbourhoods.

## 3. Data Cleaning and Preparation

### Introduction

Based on the initial exploration of the dataset, the following data cleaning and preprocessing steps were carried out on a working copy of the dataset:

- Renamed columns to provide descriptive English labels.
- Converted relevant columns to appropriate numerical data types and reformat values to standard decimal notation.
- Removed city and district-level data, retained only neighbourhood-level observations.
- Filtered the dataset to include only 2025 data.
- Restructured the data so that each neighbourhood represents one row and each feature represents one column.
- Retained only the indicators required for the 15 selected features. Where a feature required calculation from multiple indicators, the necessary component indicators were temporarily retained, combined to create the feature, and then removed.



### Import Libraries

In [1]:
# Import the Pandas library for data manipulation and analysis
import pandas as pd

### Load Dataset

In [2]:
# Loading Dataset 1: Panel of Sociodemographic Indicators of Districts and Neighborhoods of Madrid
indicators = pd.read_csv("../data/indicators.csv",sep=";")

In [3]:
# Create a working copy of dataset
indicators_2025 = indicators.copy()

In [4]:
# Display the first five rows
indicators_2025.head()

,Orden,Periodo panel,ciudad,cod_distrito,distrito,cod_barrio,barrio,año,fecha_indicador,fuente_indicador,categoría_1,categoría_2,indicador_nivel1,indicador_nivel2,indicador_nivel3,unidad_indicador,indicador_completo,valor_indicador
0,1,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Características Generales del Distrito-Barrio,NaN,Superficie,NaN,NaN,Ha,Superficie (ha.),523
1,2,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Características Generales del Distrito-Barrio,NaN,Densidad de población,NaN,NaN,Habitantes/Ha.,Población densidad (hab./ha.),258
2,3,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,NaN,NaN,Habitantes,Número habitantes,134881
3,4,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Hombres,NaN,Habitantes,Población hombres,67829
4,5,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Mujeres,NaN,Habitantes,Población mujeres,67052


### Rename columns to provide descriptive English labels

In [5]:
# Rename columns

indicators_2025.rename(columns={
    "Orden": "row_no",
    "Periodo panel": "dataset_version",
    "ciudad": "city",
    "cod_distrito": "district_code",
    "distrito": "district_name",
    "cod_barrio": "neighbourhood_code",
    "barrio": "neighbourhood_name",
    "año": "year",
    "fecha_indicador": "date",
    "fuente_indicador": "data_source",
    "categoría_1": "main_category",
    "categoría_2": "subcategory",
    "indicador_nivel1": "indicator_level1",
    "indicador_nivel2": "indicator_level2",
    "indicador_nivel3": "indicator_level3",
    "unidad_indicador": "unit",
    "indicador_completo": "indicator_name",
    "valor_indicador": "indicator_value"
}, inplace=True)


In [6]:
# Display the first ten rows following column rename
indicators_2025.head(10)

,row_no,dataset_version,city,district_code,district_name,neighbourhood_code,neighbourhood_name,year,date,data_source,main_category,subcategory,indicator_level1,indicator_level2,indicator_level3,unit,indicator_name,indicator_value
0,1,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Características Generales del Distrito-Barrio,NaN,Superficie,NaN,NaN,Ha,Superficie (ha.),523
1,2,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Características Generales del Distrito-Barrio,NaN,Densidad de población,NaN,NaN,Habitantes/Ha.,Población densidad (hab./ha.),258
2,3,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,NaN,NaN,Habitantes,Número habitantes,134881
3,4,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Hombres,NaN,Habitantes,Población hombres,67829
4,5,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Mujeres,NaN,Habitantes,Población mujeres,67052
5,6,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Edad Media,NaN,NaN,Edad,Edad media de la población,"44,03"
6,7,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,0-14,NaN,Personas,Población de 0 a 14 años,10554
7,8,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,15-29,NaN,Personas,Población de 15 a 29 años,21800
8,9,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,30-44,NaN,Personas,Población de 30 a 44 años,42857
9,10,2020,Ciudad de Madrid,1.0,Centro,NaN,NaN,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,45-64,NaN,Personas,Población de 45 a 64 años,38025


In [7]:
# Check number of shapes and rows 
indicators_2025.shape

(116908, 18)

In [8]:
# Display the number of unique dataset names
indicators_2025["neighbourhood_name"].nunique()

131

### Remove rows representing the City of Madrid and district-level data, retaining only neighbourhood-level observations

In [9]:
# Remove rows where neighbourhood_name is missing
indicators_2025= indicators_2025.dropna(subset=["neighbourhood_name"])

In [10]:
# Check number of shapes and rows 
indicators_2025.shape

(84570, 18)

In [11]:
# Check a sample 
indicators_2025.sample(10)

,row_no,dataset_version,city,district_code,district_name,neighbourhood_code,neighbourhood_name,year,date,data_source,main_category,subcategory,indicator_level1,indicator_level2,indicator_level3,unit,indicator_name,indicator_value
22292,22293,2021,Ciudad de Madrid,5.0,Chamartín,54.0,Hispanoamérica,2020,01/01/2020,Estadística Ayuntamiento de Madrid,Educación,Población en etapas educativas,Número de personas,NaN,NaN,Personas,Población en etapas educativas,4740
44682,44683,2022,Ciudad de Madrid,9.0,Moncloa-Aravaca,93.0,Ciudad Universitaria,2021,01/01/2021,Estadística Ayuntamiento de Madrid,Educación,Población en etapas educativas,Número de personas,NaN,Niñas,Personas,Etapas educativas. total niñas,1032
109602,109603,2025,Ciudad de Madrid,15.0,Ciudad Lineal,151.0,Ventas,2025,01/01/2025,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,Mujeres,NaN,Habitantes,Población mujeres,28724
73540,73541,2023,Ciudad de Madrid,20.0,San Blas-Canillejas,201.0,Simancas,2023,28/05/2023,Estadística Ayuntamiento de Madrid,Resultados Elecciones Locales,NaN,Número de votos,Partido Socialista Obrero Español,Municipio,Votos,"Elecciones municipales, votos al Partido Socia...",2517
101999,102000,2025,Ciudad de Madrid,7.0,Chamberí,71.0,Gaztambide,2025,01/01/2025,Portal de datos abiertos del Ayuntamiento de M...,Participación ciudadana,Entidades de participación ciudadana,Número de entidades,Fundaciones,NaN,Entidades,Fundaciones (sección 2ª),3
115025,115026,2025,Ciudad de Madrid,21.0,Barajas,213.0,Casco Histórico de Barajas,2023,28/05/2023,Estadística Ayuntamiento de Madrid,Resultados Elecciones Locales,NaN,Número de votos,Más Madrid,Municipio,Votos,"Elecciones municipales, votos a Más Madrid-Ver...",517
13094,13095,2020,Ciudad de Madrid,16.0,Hortaleza,162.0,Piovera,2019,01/01/2019,Estadística Ayuntamiento de Madrid,Población del distrito,Estructura de población,Número de Habitantes,65 y más,NaN,Personas,Población de 65 años y más,2066
30063,30064,2021,Ciudad de Madrid,15.0,Ciudad Lineal,151.0,Ventas,2021,01/02/2021,Estadística Ayuntamiento de Madrid,Indicadores Población desempleo,Indicadores Desempleo,Número de personas paradas,Mujeres,NaN,Personas,Paro registrado (número de personas registrada...,1831
56747,56748,2022,Ciudad de Madrid,8.0,Fuencarral-El Pardo,86.0,Valverde,2022,01/01/2022,Portal de datos abiertos del Ayuntamiento de M...,Empresa y comercio,Total locales,Locales,Total,NaN,Número,Número de locales dados de alta totales,2252
18552,18553,2020,Ciudad de Madrid,15.0,Ciudad Lineal,156.0,San Juan Bautista,2020,01/01/2020,Portal de datos abiertos del Ayuntamiento de M...,Empresa y comercio,Total locales,Locales,Total,NaN,Número,Número de locales dados de alta totales,432


In [12]:
# Display number of unique dataset names
indicators_2025["neighbourhood_name"].nunique()

131

### Filter the dataset to include only 2025 data

In [13]:
# Filter DF to only include 2025 

indicators_2025 = indicators_2025.loc[
    (indicators_2025['dataset_version'] == 2025)]
indicators_2025.head


<bound method NDFrame.head of         row_no  dataset_version              city  district_code  \
96528    96529             2025  Ciudad de Madrid            1.0   
96529    96530             2025  Ciudad de Madrid            1.0   
96530    96531             2025  Ciudad de Madrid            1.0   
96531    96532             2025  Ciudad de Madrid            1.0   
96532    96533             2025  Ciudad de Madrid            1.0   
...        ...              ...               ...            ...   
116789  116790             2025  Ciudad de Madrid           21.0   
116790  116791             2025  Ciudad de Madrid           21.0   
116791  116792             2025  Ciudad de Madrid           21.0   
116792  116793             2025  Ciudad de Madrid           21.0   
116793  116794             2025  Ciudad de Madrid           21.0   

       district_name  neighbourhood_code neighbourhood_name  year        date  \
96528         Centro                11.0            Palacio  2025  01/01

In [14]:
# Confirm the filter worked
indicators_2025['dataset_version'].nunique()

1

In [15]:
# Count missing values 

indicators_2025.isnull().sum()

row_no                   0
dataset_version          0
city                     0
district_code            0
district_name            0
neighbourhood_code       0
neighbourhood_name       0
year                     0
date                     0
data_source              0
main_category            0
subcategory           2358
indicator_level1         0
indicator_level2      1834
indicator_level3      8785
unit                     0
indicator_name           0
indicator_value          0
dtype: int64

In [16]:
# Display dimensions of the datatset
indicators_2025.shape

(14933, 18)

### Observations
Although the dataset has been filtered to include only 2025 data and observations where a neighbourhood name is present, its structure still results in a large number of rows and columns. The missing value check using isnull().sum() shows that some subcategory and indicator fields contain missing values.

For this project, indicator_name is the most relevant field for identifying and selecting socioeconomic features for the final dataset. Therefore, columns that are not required for feature selection or subsequent analysis will be dropped at this stage. 

In [17]:
# Select relevant columns
indicators_features_2025 = indicators_2025[['district_name','neighbourhood_name','main_category','indicator_name','indicator_value','unit']]

In [18]:
# Display first 5 rows 

indicators_features_2025.head()

,district_name,neighbourhood_name,main_category,indicator_name,indicator_value,unit
96528,Centro,Palacio,Características Generales del Distrito-Barrio,Superficie (ha.),"146,99",Ha
96529,Centro,Palacio,Características Generales del Distrito-Barrio,Población densidad (hab./ha.),"163,2832165",Habitantes/Ha.
96530,Centro,Palacio,Población del distrito,Número habitantes,24001,Habitantes
96531,Centro,Palacio,Población del distrito,Población hombres,11693,Habitantes
96532,Centro,Palacio,Población del distrito,Población mujeres,12308,Habitantes


In [19]:
# Sort by neighbourhood name in descending order

indicators_features_2025 = indicators_features_2025.sort_values(by="neighbourhood_name", ascending=False)

In [20]:
# Display first 5 rows 

indicators_features_2025.head()

,district_name,neighbourhood_name,main_category,indicator_name,indicator_value,unit
112048,Villaverde,Ángeles,Características Generales del Distrito-Barrio,Superficie (ha.),"193,53",Ha
112049,Villaverde,Ángeles,Características Generales del Distrito-Barrio,Población densidad (hab./ha.),"195,3288896",Habitantes/Ha.
112050,Villaverde,Ángeles,Población del distrito,Número habitantes,37802,Habitantes
112051,Villaverde,Ángeles,Población del distrito,Población hombres,17691,Habitantes
112052,Villaverde,Ángeles,Población del distrito,Población mujeres,20111,Habitantes


In [21]:
# Create a list of required feature names

selected_features = [
    "Población densidad (hab./ha.)",
    "Población mayor/igual de 25 años con bachiller elemental, graduado escolar, ESO, formación profesional 1º grado",
    "Población mayor/igual de 25 años con enseñanza primaria incompleta",
    "Población mayor/igual de 25 años con estudios superiores, licenciatura, arquitectura, ingeniería sup., estudios sup. no universitarios, doctorado, postgraduado",
    "Población mayor/igual de 25 años con formación profesional 2º grado, bachiller superior o BUP",
    "Población mayor/igual de 25 años con nivel de estudios desconocido y/o no consta",
    "Población mayor/igual de 25 años con titulación media, diplomatura, arquitectura o ingeniería técnica",
    "Población mayor/igual de 25 años que no sabe leer ni escribir o sin estudios",
    "Número de locales dados de alta en uso como vivienda",
    "Número de locales dados de alta totales",
    "Renta media disponible por persona",
    "Renta neta media anual de los hogares (Urban Audit)",
    "Tasa absoluta de paro registrado (febrero)",
    "Tasa de crecimiento demográfico (porcentaje)",
    "Tamaño medio del hogar",
    "Edad media de la población",
    "Porcentaje de envejecimiento (población mayor de 65 años/población total)",
    "Índice de vulnerabilidad bienestar social e igualdad",
    "Índice de vulnerabilidad economía y empleo",
    "Índice de vulnerabilidad educación y cultura",
    "Índice de vulnerabilidad medio ambiente urbano y movilidad",
    "Índice de vulnerabilidad salud"
]

English translation (Source: Google translate)

selected_features = [
    "Population density (inhabitants/ha)",
    "Population aged 25+ with lower secondary education (Bachiller elemental, Graduado Escolar, ESO, or Level 1 vocational training)",
    "Population aged 25+ with incomplete primary education",
    "Population aged 25+ with higher education (university degree, architecture, higher engineering, non-university higher studies, doctorate, or postgraduate studies)",
    "Population aged 25+ with Level 2 vocational training or upper secondary education (Bachiller Superior or BUP)",
    "Population aged 25+ with unknown or unrecorded education level",
    "Population aged 25+ with intermediate-level qualifications (diploma, technical architecture, or technical engineering)",
    "Population aged 25+ who are illiterate or have no formal education",
    "Number of registered premises used as housing",
    "Total number of registered premises",
    "Average disposable income per person",
    "Average annual net household income (Urban Audit)",
    "Absolute registered unemployment rate (February)",
    "Population growth rate (percentage)",
    "Average household size",
    "Average population age",
    "Aging rate (population aged 65+ / total population)",
    "Social welfare and equality vulnerability index",
    "Economy and employment vulnerability index",
    "Education and culture vulnerability index",
    "Urban environment and mobility vulnerability index",
    "Health vulnerability index"
]

In [22]:
# Check if column contains required value found in selected features

indicators_features_2025["indicator_name"].isin(selected_features)

112048    False
112049     True
112050    False
112051    False
112052    False
          ...  
116115     True
116514     True
116515    False
116516    False
116517     True
Name: indicator_name, Length: 14933, dtype: bool

In [23]:
# Keep data only where condition is true 

filtered_features_2025 = indicators_features_2025.loc[
    indicators_features_2025["indicator_name"].isin(selected_features)
]

### Observations
After selecting the relevant columns (district_name, neighbourhood_name, main_category, indicator_name, indicator_value and unit), a list was created containing the indicators required to produce the 15 selected features. The dataset was then filtered to retain only these indicators. The filtered data was previewed and its shape and data types were checked.

In [24]:
# View first 5 rows
filtered_features_2025.head()

,district_name,neighbourhood_name,main_category,indicator_name,indicator_value,unit
112049,Villaverde,Ángeles,Características Generales del Distrito-Barrio,Población densidad (hab./ha.),"195,3288896",Habitantes/Ha.
112053,Villaverde,Ángeles,Población del distrito,Edad media de la población,"45,28585601",Edad
112062,Villaverde,Ángeles,Población del distrito,Porcentaje de envejecimiento (población mayor ...,"22,28413013",Proporción
112069,Villaverde,Ángeles,Población del distrito,Tamaño medio del hogar,"2,615926058",Media
112075,Villaverde,Ángeles,Población del distrito,Tasa de crecimiento demográfico (porcentaje),"-5,490086568",Tasa


In [25]:
# Check dimensions
filtered_features_2025.shape

(2882, 6)

In [26]:
# Check data types
filtered_features_2025.dtypes

district_name         str
neighbourhood_name    str
main_category         str
indicator_name        str
indicator_value       str
unit                  str
dtype: object

### Reformatting data presentation

In [27]:
# Pivot table so that each row is one neighbourhood
pivoted_features=filtered_features_2025.pivot(index="neighbourhood_name",columns="indicator_name",values="indicator_value")

### Observations 

As each indicator is repeated for every neighbourhood and stacked vertically, the data can be organised more efficiently. The data was pivoted so that each neighbourhood represents one row, indicator names form the columns, and their corresponding values populate the dataset. As a result, district_name was removed at this stage. The reshaped data was then previewed and it was checked for missing values and for its shape.

In [28]:
# View first 5 rows
pivoted_features.head()

indicator_name,Edad media de la población,Número de locales dados de alta en uso como vivienda,Número de locales dados de alta totales,Población densidad (hab./ha.),"Población mayor/igual de 25 años con bachiller elemental, graduado escolar, ESO, formación profesional 1º grado",Población mayor/igual de 25 años con enseñanza primaria incompleta,"Población mayor/igual de 25 años con estudios superiores, licenciatura, arquitectura, ingeniería sup., estudios sup. no universitarios, doctorado, postgraduado","Población mayor/igual de 25 años con formación profesional 2º grado, bachiller superior o BUP",Población mayor/igual de 25 años con nivel de estudios desconocido y/o no consta,"Población mayor/igual de 25 años con titulación media, diplomatura, arquitectura o ingeniería técnica",...,Renta media disponible por persona,Renta neta media anual de los hogares (Urban Audit),Tamaño medio del hogar,Tasa absoluta de paro registrado (febrero),Tasa de crecimiento demográfico (porcentaje),Índice de vulnerabilidad bienestar social e igualdad,Índice de vulnerabilidad economía y empleo,Índice de vulnerabilidad educación y cultura,Índice de vulnerabilidad medio ambiente urbano y movilidad,Índice de vulnerabilidad salud
neighbourhood_name,,,,,,,,,,,,,,,,,,,,,
Abrantes,"42,52699209",54,886,"217,8128595",9674,3363,3558,5361,0,1738,...,12515,"33622,23","2,854972968","4,75","-4,858374384","7,435267188","10,6795205","13,15590765","5,137581236","6,517903051"
Acacias,"47,95211789",28,1353,"340,4695361",4914,1921,13551,5891,0,3525,...,24444,"52933,24","2,269051322","3,36","-1,119320688","6,838956625","4,9848785","6,82011066","6,257348237","6,147338816"
Adelfas,"45,79960578",8,554,"298,7203496",2472,873,6787,2942,0,1847,...,24543,"53793,35","2,332924586","2,83","-1,019626489","6,368323563","5,089996","3,952607996","6,197105237","6,897477647"
Aeropuerto,"42,31375064",11,815,"0,661050602",596,197,176,328,0,81,...,12420,30541,"2,71448468","3,99","-2,440924435","8,051756063","8,5877015","8,20500377","3,575716927","17,30113977"
Alameda de Osuna,"46,59438994",14,491,"98,75941515",2323,634,6276,3502,0,2135,...,24417,60255,"2,601780338","2,43","-0,960566229","7,385541813","4,3198675","3,691304996","5,384811544","6,682053148"


In [29]:
# Check dimensions
pivoted_features.shape

(131, 22)

In [30]:
# Count missing values 

pivoted_features.isnull().sum()

indicator_name
Edad media de la población                                                                                                                                         0
Número de locales dados de alta en uso como vivienda                                                                                                               0
Número de locales dados de alta totales                                                                                                                            0
Población densidad (hab./ha.)                                                                                                                                      0
Población mayor/igual de 25 años con bachiller elemental, graduado escolar, ESO, formación profesional 1º grado                                                    0
Población mayor/igual de 25 años con enseñanza primaria incompleta                                                                                              

### Rename columns to provide descriptive English labels

In [31]:
# Rename columns 

pivoted_features.rename(columns={

    "Edad media de la población":
        "average_population_age",

    "Número de locales dados de alta en uso como vivienda":
        "registered_residential_properties",

    "Número de locales dados de alta totales":
        "total_registered_properties",

    "Población densidad (hab./ha.)":
        "population_density",

    "Población mayor/igual de 25 años con bachiller elemental, graduado escolar, ESO, formación profesional 1º grado":
        "population_25_plus_lower_secondary_education",

    "Población mayor/igual de 25 años con enseñanza primaria incompleta":
        "population_25_plus_incomplete_primary_education",

    "Población mayor/igual de 25 años con estudios superiores, licenciatura, arquitectura, ingeniería sup., estudios sup. no universitarios, doctorado, postgraduado":
        "population_25_plus_higher_education",

    "Población mayor/igual de 25 años con formación profesional 2º grado, bachiller superior o BUP":
        "population_25_plus_upper_secondary_education",

    "Población mayor/igual de 25 años con nivel de estudios desconocido y/o no consta":
        "population_25_plus_unknown_education_level",

    "Población mayor/igual de 25 años con titulación media, diplomatura, arquitectura o ingeniería técnica":
        "population_25_plus_technical_diploma",

    "Población mayor/igual de 25 años que no sabe leer ni escribir o sin estudios":
        "population_25_plus_no_formal_education",

    "Porcentaje de envejecimiento (población mayor de 65 años/población total)":
        "ageing_rate",

    "Renta media disponible por persona":
        "average_disposable_income_per_person",

    "Renta neta media anual de los hogares (Urban Audit)":
        "average_annual_net_household_income",

    "Tamaño medio del hogar":
        "average_household_size",

    "Tasa absoluta de paro registrado (febrero)":
        "registered_unemployment_rate",

    "Tasa de crecimiento demográfico (porcentaje)":
        "population_growth_rate",

    "Índice de vulnerabilidad bienestar social e igualdad":
        "social_welfare_and_equality_vulnerability_index",

    "Índice de vulnerabilidad economía y empleo":
        "economy_and_employment_vulnerability_index",

    "Índice de vulnerabilidad educación y cultura":
        "education_and_culture_vulnerability_index",

    "Índice de vulnerabilidad medio ambiente urbano y movilidad":
        "urban_environment_and_mobility_vulnerability_index",

    "Índice de vulnerabilidad salud":
        "health_vulnerability_index"

}, inplace=True)

In [32]:
# View first 5 rows
pivoted_features.head()

indicator_name,average_population_age,registered_residential_properties,total_registered_properties,population_density,population_25_plus_lower_secondary_education,population_25_plus_incomplete_primary_education,population_25_plus_higher_education,population_25_plus_upper_secondary_education,population_25_plus_unknown_education_level,population_25_plus_technical_diploma,...,average_disposable_income_per_person,average_annual_net_household_income,average_household_size,registered_unemployment_rate,population_growth_rate,social_welfare_and_equality_vulnerability_index,economy_and_employment_vulnerability_index,education_and_culture_vulnerability_index,urban_environment_and_mobility_vulnerability_index,health_vulnerability_index
neighbourhood_name,,,,,,,,,,,,,,,,,,,,,
Abrantes,"42,52699209",54,886,"217,8128595",9674,3363,3558,5361,0,1738,...,12515,"33622,23","2,854972968","4,75","-4,858374384","7,435267188","10,6795205","13,15590765","5,137581236","6,517903051"
Acacias,"47,95211789",28,1353,"340,4695361",4914,1921,13551,5891,0,3525,...,24444,"52933,24","2,269051322","3,36","-1,119320688","6,838956625","4,9848785","6,82011066","6,257348237","6,147338816"
Adelfas,"45,79960578",8,554,"298,7203496",2472,873,6787,2942,0,1847,...,24543,"53793,35","2,332924586","2,83","-1,019626489","6,368323563","5,089996","3,952607996","6,197105237","6,897477647"
Aeropuerto,"42,31375064",11,815,"0,661050602",596,197,176,328,0,81,...,12420,30541,"2,71448468","3,99","-2,440924435","8,051756063","8,5877015","8,20500377","3,575716927","17,30113977"
Alameda de Osuna,"46,59438994",14,491,"98,75941515",2323,634,6276,3502,0,2135,...,24417,60255,"2,601780338","2,43","-0,960566229","7,385541813","4,3198675","3,691304996","5,384811544","6,682053148"


In [33]:
# Check columns

pivoted_features.columns

Index(['average_population_age', 'registered_residential_properties',
       'total_registered_properties', 'population_density',
       'population_25_plus_lower_secondary_education',
       'population_25_plus_incomplete_primary_education',
       'population_25_plus_higher_education',
       'population_25_plus_upper_secondary_education',
       'population_25_plus_unknown_education_level',
       'population_25_plus_technical_diploma',
       'population_25_plus_no_formal_education', 'ageing_rate',
       'average_disposable_income_per_person',
       'average_annual_net_household_income', 'average_household_size',
       'registered_unemployment_rate', 'population_growth_rate',
       'social_welfare_and_equality_vulnerability_index',
       'economy_and_employment_vulnerability_index',
       'education_and_culture_vulnerability_index',
       'urban_environment_and_mobility_vulnerability_index',
       'health_vulnerability_index'],
      dtype='str', name='indicator_name')

### Observations

Following the column renaming, the dataset contains the expected 22 columns and data for 131 neighbourhoods. The data is now well structured and ready for further processing to create the 15 selected features and reformat their values for modelling.

In [34]:
# Sort data alphabetically by neighbourhood name

pivoted_features=pivoted_features.sort_values('neighbourhood_name')

### Format values to standard decimal notation and into a numerical data type

In [35]:
# Format the values: remove the full stops and replace commas with decimal point

pivoted_features = pivoted_features.replace(r"\.", "", regex=True)
pivoted_features = pivoted_features.replace(",", ".", regex=True)

In [36]:
# Change values to float 

pivoted_features=pivoted_features.astype(float)

In [37]:
# View first 5 rows
pivoted_features.head()

indicator_name,average_population_age,registered_residential_properties,total_registered_properties,population_density,population_25_plus_lower_secondary_education,population_25_plus_incomplete_primary_education,population_25_plus_higher_education,population_25_plus_upper_secondary_education,population_25_plus_unknown_education_level,population_25_plus_technical_diploma,...,average_disposable_income_per_person,average_annual_net_household_income,average_household_size,registered_unemployment_rate,population_growth_rate,social_welfare_and_equality_vulnerability_index,economy_and_employment_vulnerability_index,education_and_culture_vulnerability_index,urban_environment_and_mobility_vulnerability_index,health_vulnerability_index
neighbourhood_name,,,,,,,,,,,,,,,,,,,,,
Abrantes,42.526992,54.0,886.0,217.812860,9674.0,3363.0,3558.0,5361.0,0.0,1738.0,...,12515.0,33622.23,2.854973,4.75,-4.858374,7.435267,10.679521,13.155908,5.137581,6.517903
Acacias,47.952118,28.0,1353.0,340.469536,4914.0,1921.0,13551.0,5891.0,0.0,3525.0,...,24444.0,52933.24,2.269051,3.36,-1.119321,6.838957,4.984878,6.820111,6.257348,6.147339
Adelfas,45.799606,8.0,554.0,298.720350,2472.0,873.0,6787.0,2942.0,0.0,1847.0,...,24543.0,53793.35,2.332925,2.83,-1.019626,6.368324,5.089996,3.952608,6.197105,6.897478
Aeropuerto,42.313751,11.0,815.0,0.661051,596.0,197.0,176.0,328.0,0.0,81.0,...,12420.0,30541.00,2.714485,3.99,-2.440924,8.051756,8.587701,8.205004,3.575717,17.301140
Alameda de Osuna,46.594390,14.0,491.0,98.759415,2323.0,634.0,6276.0,3502.0,0.0,2135.0,...,24417.0,60255.00,2.601780,2.43,-0.960566,7.385542,4.319867,3.691305,5.384812,6.682053


In [38]:
# Check columns

pivoted_features.columns

Index(['average_population_age', 'registered_residential_properties',
       'total_registered_properties', 'population_density',
       'population_25_plus_lower_secondary_education',
       'population_25_plus_incomplete_primary_education',
       'population_25_plus_higher_education',
       'population_25_plus_upper_secondary_education',
       'population_25_plus_unknown_education_level',
       'population_25_plus_technical_diploma',
       'population_25_plus_no_formal_education', 'ageing_rate',
       'average_disposable_income_per_person',
       'average_annual_net_household_income', 'average_household_size',
       'registered_unemployment_rate', 'population_growth_rate',
       'social_welfare_and_equality_vulnerability_index',
       'economy_and_employment_vulnerability_index',
       'education_and_culture_vulnerability_index',
       'urban_environment_and_mobility_vulnerability_index',
       'health_vulnerability_index'],
      dtype='str', name='indicator_name')

In [39]:
# Reset index 
pivoted_features = pivoted_features.reset_index()

### Observations 

Following conversion of the values to floats in standard numerical format, the index was reset, turning neighbourhood_name back into a regular column and creating a new numerical index. Keeping neighbourhood_name as the index during conversion allowed the remaining columns to be converted more easily.

In [40]:
# Confirm data types as float and check columns 
pivoted_features.dtypes

indicator_name
neighbourhood_name                                        str
average_population_age                                float64
registered_residential_properties                     float64
total_registered_properties                           float64
population_density                                    float64
population_25_plus_lower_secondary_education          float64
population_25_plus_incomplete_primary_education       float64
population_25_plus_higher_education                   float64
population_25_plus_upper_secondary_education          float64
population_25_plus_unknown_education_level            float64
population_25_plus_technical_diploma                  float64
population_25_plus_no_formal_education                float64
ageing_rate                                           float64
average_disposable_income_per_person                  float64
average_annual_net_household_income                   float64
average_household_size                                f

### Obtaining the 15 socioeconomic indicators

In [41]:
# Create the two new features

# Percentage of registered housing over total registered premises *
pivoted_features['percentage_residential_properties']= pivoted_features['registered_residential_properties']/pivoted_features['total_registered_properties']*100

# Percentage of population aged 25+ with higher education (university degree, postgraduate or doctorate) or diploma/technical degree *
pivoted_features['percentage_of_population_25_plus_either_higher_ed_or_technical']= (pivoted_features['population_25_plus_higher_education']+ pivoted_features['population_25_plus_technical_diploma'])*100/ (pivoted_features['population_25_plus_lower_secondary_education']+ pivoted_features['population_25_plus_incomplete_primary_education']+pivoted_features['population_25_plus_higher_education']+pivoted_features['population_25_plus_upper_secondary_education']+pivoted_features['population_25_plus_unknown_education_level']+pivoted_features['population_25_plus_technical_diploma']+pivoted_features['population_25_plus_no_formal_education'])


In [42]:
# Confirm column names and datatypes
pivoted_features.dtypes

indicator_name
neighbourhood_name                                                    str
average_population_age                                            float64
registered_residential_properties                                 float64
total_registered_properties                                       float64
population_density                                                float64
population_25_plus_lower_secondary_education                      float64
population_25_plus_incomplete_primary_education                   float64
population_25_plus_higher_education                               float64
population_25_plus_upper_secondary_education                      float64
population_25_plus_unknown_education_level                        float64
population_25_plus_technical_diploma                              float64
population_25_plus_no_formal_education                            float64
ageing_rate                                                       float64
average_disposable_inco

In [43]:
# Drop columns that are no longer needed
features_2025_prepared= pivoted_features.drop(columns=["population_25_plus_lower_secondary_education",
                                                          "population_25_plus_incomplete_primary_education",
                                                          "population_25_plus_higher_education",
                                                          "population_25_plus_upper_secondary_education",
                                                          "population_25_plus_unknown_education_level",
                                                          "population_25_plus_technical_diploma",
                                                          "population_25_plus_no_formal_education",
                                                          "registered_residential_properties",
                                                          "total_registered_properties"])
                                                          
                                                          

In [44]:
# Remove the label for columns left from pivot

features_2025_prepared.columns.name = None

In [45]:
# Confirm column names and datatypes
features_2025_prepared.dtypes

neighbourhood_name                                                    str
average_population_age                                            float64
population_density                                                float64
ageing_rate                                                       float64
average_disposable_income_per_person                              float64
average_annual_net_household_income                               float64
average_household_size                                            float64
registered_unemployment_rate                                      float64
population_growth_rate                                            float64
social_welfare_and_equality_vulnerability_index                   float64
economy_and_employment_vulnerability_index                        float64
education_and_culture_vulnerability_index                         float64
urban_environment_and_mobility_vulnerability_index                float64
health_vulnerability_index            

In [46]:
# View first 10 rows

features_2025_prepared.head(10)

,neighbourhood_name,average_population_age,population_density,ageing_rate,average_disposable_income_per_person,average_annual_net_household_income,average_household_size,registered_unemployment_rate,population_growth_rate,social_welfare_and_equality_vulnerability_index,economy_and_employment_vulnerability_index,education_and_culture_vulnerability_index,urban_environment_and_mobility_vulnerability_index,health_vulnerability_index,percentage_residential_properties,percentage_of_population_25_plus_either_higher_ed_or_technical
0,Abrantes,42.526992,217.812860,16.763353,12515.0,33622.23,2.854973,4.75,-4.858374,7.435267,10.679521,13.155908,5.137581,6.517903,6.094808,21.360868
1,Acacias,47.952118,340.469536,26.544208,24444.0,52933.24,2.269051,3.36,-1.119321,6.838957,4.984878,6.820111,6.257348,6.147339,2.069475,56.354576
2,Adelfas,45.799606,298.720350,23.095926,24543.0,53793.35,2.332925,2.83,-1.019626,6.368324,5.089996,3.952608,6.197105,6.897478,1.444043,56.869978
3,Aeropuerto,42.313751,0.661051,15.135967,12420.0,30541.00,2.714485,3.99,-2.440924,8.051756,8.587701,8.205004,3.575717,17.301140,1.349693,17.306397
4,Alameda de Osuna,46.594390,98.759415,26.510692,24417.0,60255.00,2.601780,2.43,-0.960566,7.385542,4.319867,3.691305,5.384812,6.682053,2.851324,55.965134
5,Almagro,46.356893,213.775456,24.439776,33425.0,78233.57,2.328984,2.08,-2.912966,8.209531,5.104565,5.823923,8.552990,8.523912,0.127173,67.589718
6,Almenara,43.541469,246.297779,18.210876,16415.0,36303.60,2.422079,4.20,-4.382039,5.784871,9.238084,9.208788,7.011487,6.958182,7.428941,36.376597
7,Almendrales,41.668011,320.407640,14.824257,12826.0,33822.43,2.819918,4.67,-4.674836,8.817065,12.138589,10.825345,7.541764,6.842716,6.474820,23.368059
8,Aluche,48.202333,244.778581,27.912372,15162.0,35834.82,2.484171,4.13,-3.518711,6.619567,8.679993,8.875167,7.129954,6.526087,9.598976,26.829224
9,Amposta,42.340086,282.009725,16.411940,10787.0,27218.54,2.684935,5.91,-6.601325,6.705769,11.712597,11.096168,4.726502,8.347488,8.478803,14.554118


In [47]:
# Dropping airport and reseting the index
features_2025_prepared = features_2025_prepared.loc[
    features_2025_prepared["neighbourhood_name"] != 'Aeropuerto'
]

features_2025_prepared = features_2025_prepared.reset_index(drop=True)

### Observations

Following the creation of the two new features, the component columns used to calculate them were dropped. The data types and resulting dataset were then checked. Aeropuerto (Airport) was also removed as it is not a residential neighbourhood, improving alignment with the model's aim of predicting residential housing prices across Madrid neighbourhoods and with the preparation of the housing price dataset. Finally, the index was reset.

In [48]:
# Preview first 10 records of cleaned dataset

features_2025_prepared.head(10)

,neighbourhood_name,average_population_age,population_density,ageing_rate,average_disposable_income_per_person,average_annual_net_household_income,average_household_size,registered_unemployment_rate,population_growth_rate,social_welfare_and_equality_vulnerability_index,economy_and_employment_vulnerability_index,education_and_culture_vulnerability_index,urban_environment_and_mobility_vulnerability_index,health_vulnerability_index,percentage_residential_properties,percentage_of_population_25_plus_either_higher_ed_or_technical
0,Abrantes,42.526992,217.812860,16.763353,12515.0,33622.23,2.854973,4.75,-4.858374,7.435267,10.679521,13.155908,5.137581,6.517903,6.094808,21.360868
1,Acacias,47.952118,340.469536,26.544208,24444.0,52933.24,2.269051,3.36,-1.119321,6.838957,4.984878,6.820111,6.257348,6.147339,2.069475,56.354576
2,Adelfas,45.799606,298.720350,23.095926,24543.0,53793.35,2.332925,2.83,-1.019626,6.368324,5.089996,3.952608,6.197105,6.897478,1.444043,56.869978
3,Alameda de Osuna,46.594390,98.759415,26.510692,24417.0,60255.00,2.601780,2.43,-0.960566,7.385542,4.319867,3.691305,5.384812,6.682053,2.851324,55.965134
4,Almagro,46.356893,213.775456,24.439776,33425.0,78233.57,2.328984,2.08,-2.912966,8.209531,5.104565,5.823923,8.552990,8.523912,0.127173,67.589718
5,Almenara,43.541469,246.297779,18.210876,16415.0,36303.60,2.422079,4.20,-4.382039,5.784871,9.238084,9.208788,7.011487,6.958182,7.428941,36.376597
6,Almendrales,41.668011,320.407640,14.824257,12826.0,33822.43,2.819918,4.67,-4.674836,8.817065,12.138589,10.825345,7.541764,6.842716,6.474820,23.368059
7,Aluche,48.202333,244.778581,27.912372,15162.0,35834.82,2.484171,4.13,-3.518711,6.619567,8.679993,8.875167,7.129954,6.526087,9.598976,26.829224
8,Amposta,42.340086,282.009725,16.411940,10787.0,27218.54,2.684935,5.91,-6.601325,6.705769,11.712597,11.096168,4.726502,8.347488,8.478803,14.554118
9,Apóstol Santiago,45.921601,132.976517,21.815427,18213.0,44215.68,2.617215,3.60,-2.602551,5.435826,6.904457,5.283034,5.493142,8.807738,9.206349,38.008574


In [49]:
# Check dimensions of cleaned dataset

features_2025_prepared.shape

(130, 16)

In [50]:
# Confirm no mising values of cleaned dataset

features_2025_prepared.isna().sum()

neighbourhood_name                                                0
average_population_age                                            0
population_density                                                0
ageing_rate                                                       0
average_disposable_income_per_person                              0
average_annual_net_household_income                               0
average_household_size                                            0
registered_unemployment_rate                                      0
population_growth_rate                                            0
social_welfare_and_equality_vulnerability_index                   0
economy_and_employment_vulnerability_index                        0
education_and_culture_vulnerability_index                         0
urban_environment_and_mobility_vulnerability_index                0
health_vulnerability_index                                        0
percentage_residential_properties               

In [51]:
# View description of cleaned dataset

features_2025_prepared.describe()

,average_population_age,population_density,ageing_rate,average_disposable_income_per_person,average_annual_net_household_income,average_household_size,registered_unemployment_rate,population_growth_rate,social_welfare_and_equality_vulnerability_index,economy_and_employment_vulnerability_index,education_and_culture_vulnerability_index,urban_environment_and_mobility_vulnerability_index,health_vulnerability_index,percentage_residential_properties,percentage_of_population_25_plus_either_higher_ed_or_technical
count,130.000000,130.000000,130.000000,130.000000,130.000000,130.000000,130.000000,130.000000,130.000000,130.000000,130.000000,130.000000,130.000000,130.000000,130.000000
mean,44.264580,196.529808,20.005912,20851.984615,49325.402615,2.525931,3.651385,-3.468187,7.630371,7.626248,7.629192,7.664802,7.536256,4.770692,44.792487
std,3.159709,127.004366,5.809450,7273.108622,18084.273179,0.293635,1.255145,2.624908,1.862743,2.771435,2.729686,6.471079,2.311860,4.371382,19.106762
min,32.852876,0.184931,3.146629,8042.000000,25584.100000,1.871816,1.580000,-27.368003,5.104424,3.884921,3.492859,2.341587,5.216657,0.000000,10.571276
25%,42.552144,92.430123,16.309994,14830.250000,34921.370000,2.315869,2.630000,-4.395248,6.081004,5.362387,5.569418,5.246387,6.517903,1.185009,27.503136
50%,44.803022,188.787995,21.031875,19373.000000,44081.100000,2.543812,3.475000,-3.094452,7.188733,6.864835,6.812861,5.984933,6.991080,3.238933,44.770274
75%,46.242733,281.856682,24.091508,26342.500000,60181.905000,2.711872,4.495000,-2.251155,8.604827,9.420169,10.042522,7.175009,7.799070,7.418720,63.250494
max,50.853616,479.203664,36.805491,36918.000000,98186.250000,3.400613,7.150000,-0.282518,14.100651,16.690711,13.712178,61.910901,22.782425,18.443396,75.079872


In [52]:
# View summary of cleaned dataset

features_2025_prepared.info()

<class 'pandas.DataFrame'>
RangeIndex: 130 entries, 0 to 129
Data columns (total 16 columns):
 #   Column                                                          Non-Null Count  Dtype  
---  ------                                                          --------------  -----  
 0   neighbourhood_name                                              130 non-null    str    
 1   average_population_age                                          130 non-null    float64
 2   population_density                                              130 non-null    float64
 3   ageing_rate                                                     130 non-null    float64
 4   average_disposable_income_per_person                            130 non-null    float64
 5   average_annual_net_household_income                             130 non-null    float64
 6   average_household_size                                          130 non-null    float64
 7   registered_unemployment_rate                                    

In [53]:
# Save prepared dataset 

features_2025_prepared.to_csv(
    "../data/features_2025_prepared.csv",
    index=False
)

### Observations

After completing the data cleaning steps, final checks confirm that the dataset contains the 15 selected features for each of the 130 neighbourhoods, with no missing values. The dataset is therefore ready to be merged with the average housing price dataset and is exported as a CSV file for the subsequent modelling stage.